pkg/cephfs/cephfs_statfs.go

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+Copyright 2018 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cephfs
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/volume"
+	"k8s.io/kubernetes/pkg/volume/util/fs"
+)
+
+const (
+	ErrCodeNoPathDefined int = iota + 1
+	ErrCodeFsInfoFailed
+)
+
+// newNoPathDefined creates a new cephfsMetricsError with code NoPathDefined.
+func newNoPathDefinedError() *cephfsMetricsError {
+	return &cephfsMetricsError{
+		Code: ErrCodeNoPathDefined,
+		Msg:  "no path defined for disk usage metrics.",
+	}
+}
+
+// newFsInfoFailedError creates a new cephfsMetricsError with code FsInfoFailed.
+func newFsInfoFailedError(err error) *cephfsMetricsError {
+	return &cephfsMetricsError{
+		Code: ErrCodeFsInfoFailed,
+		Msg:  fmt.Sprintf("Failed to get FsInfo due to error %v", err),
+	}
+}
+
+// cephfsMetricsError to distinguish different Metrics Errors.
+type cephfsMetricsError struct {
+	Code int
+	Msg  string
+}
+
+func (e *cephfsMetricsError) Error() string {
+	return e.Msg
+}
+
+var _ volume.MetricsProvider = &cephfsStatFS{}
+
+// cephfsStatFS represents a MetricsProvider that calculates the used and available
+// Volume space by stat'ing and gathering filesystem info for the Volume path.
+type cephfsStatFS struct {
+	// the directory path the volume is mounted to.
+	path string
+}
+
+// NewcephfsStatFS creates a new cephfsStatFS with the Volume path.
+func NewcephfsStatFS(path string) volume.MetricsProvider {
+	return &cephfsStatFS{path}
+}
+
+// GetMetrics calculates the volume usage and device free space by executing "du"
+// and gathering filesystem info for the Volume path.
+func (md *cephfsStatFS) GetMetrics() (*volume.Metrics, error) {
+	metrics := &volume.Metrics{Time: metav1.Now()}
+	if md.path == "" {
+		return metrics, newNoPathDefinedError()
+	}
+
+	err := md.getFsInfo(metrics)
+	if err != nil {
+		return metrics, err
+	}
+
+	return metrics, nil
+}
+
+// getFsInfo writes metrics.Capacity, metrics.Used and metrics.Available from the filesystem info
+func (md *cephfsStatFS) getFsInfo(metrics *volume.Metrics) error {
+	available, capacity, usage, inodes, inodesFree, inodesUsed, err := fs.FsInfo(md.path)
```

Comment:
@humblec (tagging @ajarr as well) The call to fs.FsInfo uses the syscall statfs. I checked that statfs on a CephFS **subvolume** mount, returns used space as per the set attrs on the subvolume. I wanted to know if the same applies to inodes, IOW, statfs  fields,
```
fsfilcnt_t   f_files;   /* total file nodes in file system */
fsfilcnt_t   f_ffree;   /* free file nodes in fs */
```
and hence can be used to generate metrics as below.

Reply:
@ShyamsundarR not sure I got the comment, I would like to mention that, whatever `statfs` returns is good enough for us. If subvol is not responding properly to `statfs` its totally a different bug which Ceph has to take care. I dont think there is a bug though.

Reply:
@humblec I tested and determined that cephfs subvolume mounts as used in ceph-csi returns the correct total space as needed by us (i.e it sticks to the quota set on the subvolume). I am however not able to determine if it returns inode count information as required  (or used space, which is a short test away if needed). Hence I need confirmation that it does, in case you know it or can test it, else from CephFS developers.

We need to understand statfs limitations **if any** for CephFS subvolumes, before interpreting it's output the way we want OR assuming any errors is a bug in CephFS subvolume implementation.

The provided image shows used space, which was never my concern as stated in the comment.

Hope this provides the required clarity.

Reply:
>We need to understand statfs limitations if any for CephFS subvolumes, before interpreting it's output the way we want OR assuming any errors is a bug in CephFS subvolume implementation.
>
@ShyamsundarR  If you mount the subvolume and call stat against, you should be able to verify it , Isnt it ?